### ISSUE TITLE
weblogCallback corrupting variables on the stack?

### ISSUE THREAD

**Issue by frankonski** [HomeSpan/HomeSpan#903]
Using weblogCallback, even if not setting any value to the string parameter, is corrupting variables on the stack (seem like it!).

I have a `static byte panelData[16]` that is declared and used by a different library (`dscKeybusInterface`) which gets modified if `homeSpan.setWeblogCallback(extraWebData)` is defined, even if `void extraWebData(String &r)` does not modify `r`.

I kept wondering why hitting F5 repeatedly on the `/status` URL of the weblogs would end up changing the `SecuritySystem` current state and target state. After further investigation, these states get only set when reading values from that `static byte panelData[16]` variable. That 3rd party library also performs checksums of messages sent to a data bus, and CRC checksum errors keeps happening **only** on a page reload of `/status`.

The only way to work around this is not not set the weblog callback at all.

Looking at `HAP.cpp`, it looks like `usrString` ends up corrupting the other variable.
```
  if(homeSpan.weblogCallback){
    String usrString;
    homeSpan.weblogCallback(usrString); // Callback to add user-defined html in top table.
    hapOut << usrString.c_str();    
  }
```

Note that I also added many calls to `heap_caps_check_integrity_all(true)` and it always return `true` / `1`.

.. I won't be able to look at this for few weeks..

**Comment by frankonski** [comment]
I got similar memory corruption by forcing a wifi disconnect by using my router's app to block, unblock, block, unblock.. the board, causing WiFi disconnects and reconnects. Just doing this ended up messing up the content of that `panelData[]` variable.

Should I give up trying to find the cause and see what happens with 2.0.0rc1?

**Comment by HomeSpan** [comment]
There's no changes to the WiFi portion of the code in 2.0.0-rc1, but I will be refactoring it all in 2.1.0.  Espressif made a lot of changes to the WiFi classes but none of it is documented.  I'm going through the Arduino -ESP32 code to try to figure out what the latest and greatest is.

**Comment by HomeSpan** [comment]
Are you using homeSpan.autoPoll() or simply calling poll() in the main loop.  If you call poll() directly in the main loop there is no threading and there should be no conflicts unless your dscKeybusInterface library is separately threaded.  See if that makes a difference.  If the issue does only show up when using autoPoll() it is likely a threading conflict - not quite sure how to solve.  Can you provide a bare-bones sketch the illustrates the problem so I can explore a solution?

**Comment by frankonski** [comment]
I have been calling `poll()` for quite some times as using `autoPoll()` seemed to not call the `poll()` method of all the accessories often enough, and was causing some "No responses"... That was long ago, before I started annoying you with issues & PR. I could give it a try again to gather more details now that I am more familiar with HS code base.

And without hooking an ESP32 board to a DSC alarm system I wouldn't know how to reproduce the issue...

**Comment by HomeSpan** [comment]
Can you send me a sample of the code in your weblog that causes the issue?

**Comment by HomeSpan** [comment]
Understood, but can you isolate the problem somehow.  For example, if you toss everything from your sketch except a static variable, do you see the problem?  Also, are you using the normal HomeSpan call to the weblog or are you using your TLS code.  I'm not sure how to debug without a case we can investigate in detail.    Also, does this occur in any other sketch

It may be that you have a memory leak somewhere in your code, or there is an issue in the DSC code, which never surfaced until now in the new release.

**Comment by frankonski** [comment]
I maintain my own fork of dscKeybusInterface now, and I got inspiration from one of its example that uses HomeSpan. That example has a really flawed implementation, so I'll see if I can push upstream a better, simpler one that would show the behavior, and just force a value to that variable that would be a valid one according to a good state.

**Comment by frankonski** [comment]
All of the TLS code is still out (using `#ifdef`). I tried with a dummy sample to no avail, probably because I need to use the ISR function that reads data from the panel.

`panelData` is set in the main `loop()` at https://github.com/frankonski/dscKeybusInterface/blob/master/src/dscKeybusInterface.cpp#L147 from a `volatile` variable `panelBuffer` in an ISR function, https://github.com/frankonski/dscKeybusInterface/blob/master/src/dscKeybusInterface.cpp#L518

The issue only presents itself when I hit non-stop the browser's Refresh button. So technically, **not** hitting refresh still gets `loop()` called which processes the state of that `panelData` variable.

**Comment by frankonski** [comment]
Maybe I will try creating a task on a different core to attach the interrupt, so to have WiFi interrupts on a separate core.

**Comment by frankonski** [comment]
I have found out the hard way that `xPortGetCoreID()` on arduino-espressif 2.0.17 returns a hardcoded value of 0, so not that useful to see if this is an interrupt issue between wifi and that 3rd party ISR routine.. I will instead try to upgrade everything to HS 2.1 and arduino-espressif to 3.x to see if the behaviour remains the same. 

I am now thinking the issue might be related to timing of reading bits on a line, skipping bits that ends up corrupting the data, rather than stack corruption. I’ll also review that library’s code, as checksums should fail if bits are missing..

**Comment by frankonski** [comment]
Well, after a bit of review and testing, I can confirm this is not a stack issue, but an interrupt / synchronization problem.
I'm able to see that when I thought the stack was getting corrupted, what is really happening is that some bits are missing, or not read fast enough... and of course not all messages have checksums. The missing bits are likely caused by WiFi using interrupts (from what I read online) when using weblogs intensively, and that 3rd party library also triggering an interrupt from a 250ms timer after a clock line raises or falls from another interrupt. 
Somewhat strange how it was implemented.

Nevertheless, my next step will be to upgrade to HS 2.1 and see how everything behaves afterwards.

**Comment by HomeSpan** [comment]
Thanks for the update.  It's not inherently a problem for an interrupt to trigger other interrupts, but the code needs to be very carefully written to make sure variables are protected, etc.

One frequent issue is interrupts taking too much time when other threads are doing other stuff.  Espressif recommends the use of message queues to solve issues with overlapping data access, which you will see peppered throughout HomeSpan in various places.

**Comment by frankonski** [comment]
Yeah, my assumption right now is that one of the WiFi ISR is taking too much time, which causes the ISR of that 3rd party library to  miss reading the data line when it should. Then one bit is missed and the whole rest of the bytes are messed up.

.. and I was reading about SPI, MOSI, MISO and SCK pins of ESP32 chips which already exists, that could hopefully be used instead with the SPI library. Hopefully that would be much more reliable than the software implementation of reading the data line when the clock like is HIGH, and writing on the data line when the clock line is LOW. :-/
long term project...

**Comment by frankonski** [comment]
and for the record, I get the same error with HomeSpan 2.1.0 / Arduino-ESP 3.0.7 / ESP-IDF 5.1.4.

Since the issue has been nailed down to be interrupt-related, unrelated to HomeSpan, feel free to close if this fits you.

**Comment by HomeSpan** [comment]
okay - thanks.

**Comment by frankonski** [comment]
I can confirm this is ISR related. Hitting F5 to constantly hit the /status URL has a direct impact on the two ISR used by this 3rd party library (one ISR starts a 250µs timer that calls another ISR), and with arduino-esp32 3.1.1 I am now seeing timer-related errors logged by esp itself
```
E (8715) gptimer: gptimer_start(396): timer is not ready for a new start
```
New wifi connections must be using interupts, causing delays in the ISR, causing missed data pin readings, causing corruption.
I need to  work on this 3rd party library to fix how it works... :(
